### PR TITLE
Build against Ubuntu

### LINKED ISSUE #28: Building gbdk-lib on ubuntu fails
```
make -C libc
make[1]: Entering directory '/mnt/c/development/magazine_repo/gbdk-2020/gbdk-lib/libc'
for i in gbz80; do make port THIS=$i PORT=$i; done
make[2]: Entering directory '/mnt/c/development/magazine_repo/gbdk-2020/gbdk-lib/libc'
rm -f  *.o *.cdb *.sym *.lst *~ *.asm
if [ -e global.s ]; then \
        sed -e "s/.NEAR_CALLS\W=\W[0-9]\+/.NEAR_CALLS = 1/" global.s > tmp1.txt ;\
        mv tmp1.txt global.s; \
fi
mkdir -p ../build/small/asxxxx/gbz80
/bin/sdcc -I../include -DSDCC_z80 -D__gbz80 -mgbz80 -DGBDK=1 -c -o ../build/small/asxxxx/gbz80/bsearch.o bsearch.c
make[2]: /bin/sdcc: Command not found
../libc/rules-asxxxx.mk:7: recipe for target '../build/small/asxxxx/gbz80/bsearch.o' failed
make[2]: *** [../build/small/asxxxx/gbz80/bsearch.o] Error 127
make[2]: Leaving directory '/mnt/c/development/magazine_repo/gbdk-2020/gbdk-lib/libc'
Makefile:31: recipe for target 'ports' failed
make[1]: *** [ports] Error 2
make[1]: Leaving directory '/mnt/c/development/magazine_repo/gbdk-2020/gbdk-lib/libc'
Makefile:4: recipe for target 'all' failed
make: *** [all] Error 2
```

### LINKED ISSUE #46: Build fails when installing SDCC (last step)
On Linux Ubuntu 19.10 (64-bit), GBDK-2020 builds correctly, but the last step fails with:

```
Installing SDCC
cp: cannot stat '/usr/bin/{as2gbmap,makebin,packihx,sdar,sdasgb,sdcc,sdcdb,sdcdb{,src}.el,sdcpp,sdldgb,sdnm,sdobjcopy,sdranlib,sz80}': No such file or directory
make: *** [Makefile:213: sdcc-install] Error 1
```

I copied the command and ran it myself, and all those files exist but 4:

```
cp: cannot stat '/usr/bin/sdcdb': No such file or directory
cp: cannot stat '/usr/bin/sdcdb.el': No such file or directory
cp: cannot stat '/usr/bin/sdcdbsrc.el': No such file or directory
cp: cannot stat '/usr/bin/sz80': No such file or directory
```

Also, it seems Make is not correctly interpreting that line (`/usr/bin/{as2gbmap,makebin,packihx,sdar,sdasgb,sdcc,sdcdb,sdcdb{,src}.el,sdcpp,sdldgb,sdnm,sdobjcopy,sdranlib,sz80}`), it is feeding it literally to `cp` without expanding the curly braces to get the individual file names.

### PR DESCRIPTION
**Pull request by drpaneas**
* Modify Makefile to include Linux (based on Ubuntu).
* Update documentation with the required Ubuntu packages and conf.

Fixes #46
Fixes #28

### CODE REVIEW

**Comment by drpaneas** [comment]
@Zal0 @untoxa please review whenever you have the time :)

**Comment by drpaneas** [comment]
I've added an automated build release workflow in GitHub Actions. See an example at my fork:
- https://github.com/drpaneas/gbdk-2020/runs/1079710317?check_suite_focus=true
- https://github.com/drpaneas/gbdk-2020/releases/tag/v3.2 (the build is inside the `zip` archive)

**Comment by untoxa** [comment]
there will be a problem with upcoming version of GBDK. it will use the latest version of SDCC which is avaliable in daily builds, but not avaliable in the ubuntu repo. it is essential to use that version, because it has VERY inportant improvements, and it is not practical to wait a year or two untill that appear in the SDCC release and then will be avaliable in ubuntu. it is stable enough for GBDK-2020. but your dependencies will break.

**Comment by drpaneas** [comment]
Hi @untoxa 

Thank for the quick feedback on this.

> there will be a problem with upcoming version of GBDK. it will use the latest version of SDCC which is avaliable in daily builds, but not avaliable in the ubuntu repo.

Then, I will try building it then inside a container and expose the `/build` directory as artifact and check if this works. Which is the correct working version of `sdcc`? Can you give me a link maybe?

**Comment by untoxa** [comment]
http://sdcc.sourceforge.net/snap.php
i use rev #11845 myself. it works well, i don't see any critical issues with it.

the main problem, for now, is adapting LCC to make it work with the native SDCC toolchain, and at the same time to keep it backward-compatible. ldldgb accepts arguments in a slightly different format and -yt, -yn and some other GB-specific linker arguments must not be passed into the linker anymore, they are for makebin utility.

that turns out to be not a simple task

**Comment by drpaneas** [comment]
Ok, so I have updated the commits and now I have a pipeline where I build it based on the latest linux 64-bit from the http://sdcc.sourceforge.net/snap.php

It builds fine. But ... using it fails:

```c
drpaneas@DESKTOP-3G75NQG:~$ ./gbdk/bin/lcc -Wa-l -Wl-m -Wl-j -DUSE_SFR_FOR_REG -c -o main.o main.c
drpaneas@DESKTOP-3G75NQG:~$ ls
gbdk  gbdk.zip  github  main.c  main.lst  main.o  main.sym
drpaneas@DESKTOP-3G75NQG:~$ ./gbdk/bin/lcc -Wa-l -Wl-m -Wl-j -DUSE_SFR_FOR_REG -o main.gb main.o
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
Error: Format(XL3) Only XL2 format is supported
WARNING: possibly wrote twice at addr 0 (00->00)
WARNING: possibly wrote twice at addr 20 (00->00)
WARNING: possibly wrote twice at addr 20 (00->00)
WARNING: possibly wrote twice at addr 28 (00->00)
WARNING: possibly wrote twice at addr 40 (00->00)
WARNING: possibly wrote twice at addr 40 (00->00)
WARNING: possibly wrote twice at addr 50 (00->00)
WARNING: possibly wrote twice at addr 50 (00->00)
WARNING: possibly wrote twice at addr 58 (00->00)
WARNING: possibly wrote twice at addr 58 (00->00)
WARNING: possibly wrote twice at addr 58 (00->00)
WARNING: possibly wrote twice at addr 60 (00->00)
WARNING: possibly wrote twice at addr 60 (00->00)
WARNING: possibly wrote twice at addr 60 (00->00)
WARNING: possibly wrote twice at addr 80 (00->00)
WARNING: possibly wrote twice at addr 80 (00->00)
WARNING: possibly wrote twice at addr 82 (00->D5)
WARNING: possibly wrote twice at addr 82 (00->00)
WARNING: possibly wrote twice at addr 8f (00->F3)
WARNING: possibly wrote twice at addr 8f (00->00)
WARNING: possibly wrote twice at addr 100 (00->00)
WARNING: possibly wrote twice at addr 100 (00->00)
WARNING: possibly wrote twice at addr 10d (00->03)
WARNING: possibly wrote twice at addr 11a (00->89)
WARNING: possibly wrote twice at addr 127 (00->67)
WARNING: possibly wrote twice at addr 134 (00->3E)
WARNING: possibly wrote twice at addr 134 (00->00)
WARNING: possibly wrote twice at addr 144 (00->00)
WARNING: possibly wrote twice at addr 147 (00->00)
WARNING: possibly wrote twice at addr 147 (00->00)
WARNING: possibly wrote twice at addr 148 (00->00)
WARNING: possibly wrote twice at addr 148 (00->00)
WARNING: possibly wrote twice at addr 149 (00->00)
WARNING: possibly wrote twice at addr 149 (00->00)
WARNING: possibly wrote twice at addr 14a (00->00)
WARNING: possibly wrote twice at addr 14a (00->00)
WARNING: possibly wrote twice at addr 14c (00->00)
WARNING: possibly wrote twice at addr 14c (00->00)
WARNING: possibly wrote twice at addr 14d (00->01)
WARNING: possibly wrote twice at addr 14d (00->00)
WARNING: possibly wrote twice at addr 14e (00->00)
WARNING: possibly wrote twice at addr 14e (00->00)
WARNING: possibly wrote twice at addr 150 (00->00)
WARNING: possibly wrote twice at addr 150 (00->00)
WARNING: possibly wrote twice at addr 150 (00->00)
WARNING: possibly wrote twice at addr 15d (00->00)
WARNING: possibly wrote twice at addr 15d (00->00)
WARNING: possibly wrote twice at addr 169 (00->80)
WARNING: possibly wrote twice at addr 169 (00->00)
WARNING: possibly wrote twice at addr 176 (00->21)
WARNING: possibly wrote twice at addr 178 (00->FE)
WARNING: possibly wrote twice at addr 178 (00->00)
WARNING: possibly wrote twice at addr 185 (00->3E)
WARNING: possibly wrote twice at addr 190 (00->0A)
WARNING: possibly wrote twice at addr 190 (00->00)
WARNING: possibly wrote twice at addr 19d (00->01)
WARNING: possibly wrote twice at addr 1aa (00->1B)
WARNING: possibly wrote twice at addr 1c4 (00->E0)
WARNING: possibly wrote twice at addr 1d0 (00->0C)
WARNING: possibly wrote twice at addr 1d4 (00->00)
WARNING: possibly wrote twice at addr 1d4 (00->00)
WARNING: possibly wrote twice at addr 1d4 (00->00)
WARNING: possibly wrote twice at addr 1e0 (00->00)
ERROR: address overflow (addr c0a0 >= 8000)
```

I guess I have to fetch not the latest but the specific one you mentioned that is working for you. I will test this and let you know.

**Comment by untoxa** [comment]
definitely, that's what i'm talking about. link-gbz80 should be thrown out, sdldgb and makebin utility should be used instead. lcc utility should be refactored.
it should be built like this: https://github.com/untoxa/rhythm
lcc is a frontend, analogue of my @make.bat

**Comment by drpaneas** [comment]
oh I see ... this is ugly ... :(

Are there any plans refactoring `lcc`? This goes beyond my level of expertise, so I cannot help you with that. Is there anything else perhaps I could do to help having gbdk-2020 in Linux as well?

**Comment by untoxa** [comment]
You may write makefiles instead of using lcc. Rework examples: place them into separate folders with own makefile each. Don’t forget batch files for windows too.

**Comment by untoxa** [comment]
maybe we should throw lcc away and replace it with make, for windows too. one half of lcc is outdated/not working.

**Comment by drpaneas** [comment]
Ok, feel free to create a github issue with the acceptance criteria for this task and assign it to me ;)

**Comment by Zal0** [comment]
Hi, I have plans to refactor lcc during  the next days
The changes are not very hard, but I need to sit down and do it carefully

File: Makefile [hunk 1]
```diff
@@ -210,20 +211,25 @@ linker-clean:
 sdcc-install: check-SDCCDIR
 	@echo Installing SDCC
 	@mkdir -p $(BUILDDIR)/bin/SDCC/bin
+
 ifeq ($(OS),Windows_NT)
 		@cp -r $(SDCCDIR)/bin $(BUILDDIR)/bin/SDCC
-else
+endif
+ifeq ($(OS),Linux)
+		@cp $(SDCCDIR)/bin/* $(BUILDDIR)/bin/SDCC/bin/
```
Comment (JL2210) [inline Makefile:219]: Seriously? What if SDCCDIR=/usr? This is the same problem we had before.

**Comment by untoxa** [comment]
fixed in v4 release

**Pull request closed without merging**